src/Native/System.Native/pal_exec.cpp

```diff
@@ -0,0 +1,142 @@
+        (redirectStdout && pipe(stdoutFds) != 0) ||
+        (redirectStderr && pipe(stderrFds) != 0))
+    {
+        assert(!"pipe() failed.");
+        success = false;
+        goto done;
+    }
+
+    // Fork the child process
+    if ((processId = fork()) == -1)
+    {
+        assert(!"fork() failed.");
+        success = false;
+        goto done;
+    }
+
+    /* From the time the child process (processId == 0) begins running from fork to when
+     * it reaches execve, the child process must not touch anything in the PAL.  Doing so
+     * is not safe. The parent process (processId >= 0) may continue to use the PAL.
+     */
+
+    if (processId == 0) // processId == 0 if this is child process
+    {
+        // Close the parent end of any open pipes
+        CloseIfOpen(stdinFds[WRITE_END_OF_PIPE]);
+        CloseIfOpen(stdoutFds[READ_END_OF_PIPE]);
+        CloseIfOpen(stderrFds[READ_END_OF_PIPE]);
+
+        // For any redirections that should happen, dup the pipe descriptors onto stdin/out/err.
+        // Then close out the old pipe descriptrs, which we no longer need.
+        if ((redirectStdin  && dup2(stdinFds[READ_END_OF_PIPE],   STDIN_FILENO)  == -1) ||
+            (redirectStdout && dup2(stdoutFds[WRITE_END_OF_PIPE], STDOUT_FILENO) == -1) ||
+            (redirectStderr && dup2(stderrFds[WRITE_END_OF_PIPE], STDERR_FILENO) == -1))
+        {
+            _exit(errno != 0 ? errno : EXIT_FAILURE);
+        }
+        CloseIfOpen(stdinFds[READ_END_OF_PIPE]);
+        CloseIfOpen(stdoutFds[WRITE_END_OF_PIPE]);
+        CloseIfOpen(stderrFds[WRITE_END_OF_PIPE]);
+
+        // Change to the designated working directory, if one was specified
+        if (nullptr != cwd && chdir(cwd) == -1)
+        {
+            _exit(errno != 0 ? errno : EXIT_FAILURE);
+        }
+
+        // Finally, execute the new process.  execve will not return if it's successful.
+        execve(filename, (char**)argv, (char**)envp);
+        _exit(errno != 0 ? errno : EXIT_FAILURE); // execve failed
+    }
+
+    // This is the parent process. processId == pid of the child
+    *childPid = processId;
+    *stdinFd = stdinFds[WRITE_END_OF_PIPE];
+    *stdoutFd = stdoutFds[READ_END_OF_PIPE];
+    *stderrFd = stderrFds[READ_END_OF_PIPE];
+
+done:
+    // Regardless of success or failure, close the parent's copy of the child's end of
+    // any opened pipes.  The parent doesn't need them anymore.
+    CloseIfOpen(stdinFds[READ_END_OF_PIPE]);
+    CloseIfOpen(stdoutFds[WRITE_END_OF_PIPE]);
+    CloseIfOpen(stderrFds[WRITE_END_OF_PIPE]);
+
+    // If we failed, close everything else and give back error values in all out arguments.
+    if (!success)
+    {
+        CloseIfOpen(stdinFds[WRITE_END_OF_PIPE]);
+        CloseIfOpen(stdoutFds[READ_END_OF_PIPE]);
+        CloseIfOpen(stderrFds[READ_END_OF_PIPE]);
+
+        *stdinFd  = -1;
+        *stdoutFd = -1;
+        *stderrFd = -1;
+        *childPid = -1;
+    }
+    
```

Comment:
The intention here was that if the fork failed, errno would still be set to whatever it was from the fork call.  But I just noticed we're making intervening calls to close, and it's possible close is changing errno.  Consider capturing errno after the fork and resetting it here.